File: dal-python/tests/test_xccy_calibration.py
"""Tests for cross-currency calibration."""

import dal


def _today():
    return dal.Date_(2025, 6, 20)


def _spot():
    return _today().AddDays(2)


def _make_baseline_curves():
    """Build flat 4% OIS curves for USD and EUR."""
    knot_dates = [_spot(), _spot().AddDays(3650)]
    ois_rates = [0.04, 0.04]

    usd_ois = dal.DiscountPWLF_New("usd_ois", "USD", knot_dates, ois_rates)
    eur_ois = dal.DiscountPWLF_New("eur_ois", "EUR", knot_dates, ois_rates)

    usd_block = dal.CurveBlock_New(usd_ois, libor_basis=dal.DayBasis_New("ACT_365F"))
    eur_block = dal.CurveBlock_New(eur_ois, libor_basis=dal.DayBasis_New("ACT_360"))

    return usd_block, eur_block


# ---- CrossCurrencyCalibrationSpecBuilder_ ----

def test_xccy_builder_defaults():
    """CrossCurrencyCalibrationSpecBuilder_ has sensible defaults."""
    builder = dal.CrossCurrencyCalibrationSpecBuilder_()
    assert builder.tolerance_ == 1.0e-10  # nosec B101 - pytest assertions are intentional
    assert builder.fitTolerance_ == 1.0e-6  # nosec B101 - pytest assertions are intentional
    assert builder.maxEvaluations_ == 200  # nosec B101 - pytest assertions are intentional
    assert builder.maxRestarts_ == 20  # nosec B101 - pytest assertions are intentional
    assert builder.initialGuess_ == 0.0  # nosec B101 - pytest assertions are intentional
    assert builder.smoothingWeight_ == 1.0  # nosec B101 - pytest assertions are intentional
    assert builder.fxSpot_ == 0.0  # nosec B101 - pytest assertions are intentional


def test_xccy_builder_can_set_fields():
    """CrossCurrencyCalibrationSpecBuilder_ fields can be set."""
    builder = dal.CrossCurrencyCalibrationSpecBuilder_()
    builder.today_ = _today()
    builder.basisPair_ = dal.CurrencyPair_New("USD", "EUR")
    builder.fxSpot_ = 1.10
    builder.tolerance_ = 1.0e-8
    builder.solveMode_ = dal.CurveSolveMode.APPROXIMATE
    assert builder.fxSpot_ == 1.10  # nosec B101 - pytest assertions are intentional
    assert builder.tolerance_ == 1.0e-8  # nosec B101 - pytest assertions are intentional


def test_xccy_builder_per_node_guess_has_legacy_and_snake_case_names():
    builder = dal.CrossCurrencyCalibrationSpecBuilder_()
    builder.initial_guess_per_node = [0.0125, 0.0175]

    assert builder.initialGuessPerNode_ == [0.0125, 0.0175]  # nosec B101
    assert builder.Build().initial_guess_per_node == [0.0125, 0.0175]  # nosec B101


def test_xccy_builder_new_fields_have_legacy_and_snake_case_names():
    """Reset-aware fields do not replace the existing underscore surface."""
    builder = dal.CrossCurrencyCalibrationSpecBuilder_()
    valuation_time = dal.DateTime_(_today(), 9, 45)
    snapshot = dal.MarketFixingSnapshot_New({})

    builder.valuation_time = valuation_time
    builder.collateralCurrency_ = dal.Ccy_("USD")
    builder.fixings = snapshot
    builder.fx_spot = 1.10

    assert builder.valuationTime_ is not None  # nosec B101 - pytest assertions are intentional
    assert builder.collateral_currency is not None  # nosec B101 - pytest assertions are intentional
    assert builder.fixings_ is snapshot  # nosec B101 - pytest assertions are intentional
    assert builder.fxSpot_ == 1.10  # nosec B101 - pytest assertions are intentional


def test_legacy_xccy_constructor_accepts_every_original_positional_argument():
    """The config overload leaves the original positional order untouched."""
    currencies = dal.CurrencyPair_New("USD", "EUR")
    domestic_leg = dal.RateLegConvention_New(dal.PeriodLength_New("6M"), dal.DayBasis_New("ACT_365F"))
    domestic_index = dal.RateIndexConvention_New(
        dal.PeriodLength_New("3M"), dal.DayBasis_New("ACT_360"), dal.CollateralType_OIS()
    )
    foreign_leg = dal.RateLegConvention_New(dal.PeriodLength_New("6M"), dal.DayBasis_New("ACT_360"))
    foreign_index = dal.RateIndexConvention_New(
        dal.PeriodLength_New("6M"), dal.DayBasis_New("ACT_360"), dal.CollateralType_OIS()
    )

    instrument = dal.CrossCurrencySwap_New(
        _today(),
        _spot(),
        _spot().AddDays(365),
        0.001,
        currencies,
        125.0,
        113.5,
        domestic_leg,
        domestic_index,
        foreign_leg,
        foreign_index,
    )
    assert instrument is not None  # nosec B101 - pytest assertions are intentional


# ---- Cross-currency calibration ----

def _make_xccy_instruments(fx_spot, years=(2, 5, 10)):
    """Build cross-currency swap instruments for calibration."""
    currencies = dal.CurrencyPair_New("USD", "EUR")
    usd_leg = dal.RateLegConvention_New(dal.PeriodLength_New("6M"), dal.DayBasis_New("ACT_365F"))
    usd_index = dal.RateIndexConvention_New(
        dal.PeriodLength_New("3M"), dal.DayBasis_New("ACT_360"), dal.CollateralType_OIS()
    )
    eur_leg = dal.RateLegConvention_New(dal.PeriodLength_New("6M"), dal.DayBasis_New("ACT_360"))
    eur_index = dal.RateIndexConvention_New(
        dal.PeriodLength_New("6M"), dal.DayBasis_New("ACT_360"), dal.CollateralType_OIS()
    )

    instruments = []
    knot_dates = []
    for y in years:
        maturity = _spot().AddDays(y * 365)
        knot_dates.append(maturity)
        inst = dal.CrossCurrencySwap_New(
            _today(), _spot(), maturity, 0.01, currencies,
            domestic_notional=100.0, foreign_notional=100.0 / fx_spot,
            domestic_leg=usd_leg, domestic_index=usd_index,
            foreign_leg=eur_leg, foreign_index=eur_index,
        )
        instruments.append(inst)
    return instruments, knot_dates


def _make_xccy_spec(solve_mode, years=(2, 5, 10)):
    usd_block, eur_block = _make_baseline_curves()
    instruments, knot_dates = _make_xccy_instruments(fx_spot=1.10, years=years)

    builder = dal.CrossCurrencyCalibrationSpecBuilder_()
    builder.today_ = _today()
    builder.basisPair_ = dal.CurrencyPair_New("USD", "EUR")
    builder.domesticCurveBlock_ = usd_block
    builder.foreignCurveBlock_ = eur_block
    builder.fxSpot_ = 1.10
    builder.tolerance_ = 1.0e-8
    builder.solveMode_ = solve_mode
    builder.initialGuess_ = 0.01
    builder.instruments_ = instruments
    builder.knotDates_ = knot_dates
    return builder.Build(), knot_dates


def test_calibrate_xccy_market():
    """CalibrateXccyMarket calibrates a cross-currency basis curve."""
    usd_block, eur_block = _make_baseline_curves()
    instruments, knot_dates = _make_xccy_instruments(fx_spot=1.10)

    builder = dal.CrossCurrencyCalibrationSpecBuilder_()
    builder.today_ = _today()
    builder.basisPair_ = dal.CurrencyPair_New("USD", "EUR")
    builder.domesticCurveBlock_ = usd_block
    builder.foreignCurveBlock_ = eur_block
    builder.fxSpot_ = 1.10
    builder.tolerance_ = 1.0e-8
    builder.solveMode_ = dal.CurveSolveMode.APPROXIMATE
    builder.initialGuess_ = 0.01
    builder.instruments_ = instruments
    builder.knotDates_ = knot_dates

    spec = builder.Build()
    result = dal.CalibrateXccyMarket(spec)

    diag = result.diagnostics_
    assert len(diag.marketRates_) > 0  # nosec B101 - pytest assertions are intentional
    assert len(diag.marketRates_) == len(diag.modelRates_)  # nosec B101 - pytest assertions are intentional
    assert diag.maxAbsResidual_ < 1.0e-4  # nosec B101 - pytest assertions are intentional
    assert diag.rmsResidual_ < 1.0e-4  # nosec B101 - pytest assertions are intentional

    # FX forward curve should be populated
    fwd = result.fxForwardCurve_
    assert len(fwd.dates_) > 0  # nosec B101 - pytest assertions are intentional
    assert len(fwd.dates_) == len(fwd.forwards_)  # nosec B101 - pytest assertions are intentional


def test_calibrate_xccy_market_result_has_market():
    """CalibrateXccyMarket result contains a market and FX forward curve."""
    usd_block, eur_block = _make_baseline_curves()
    instruments, knot_dates = _make_xccy_instruments(fx_spot=1.10)

    builder = dal.CrossCurrencyCalibrationSpecBuilder_()
    builder.today_ = _today()
    builder.basisPair_ = dal.CurrencyPair_New("USD", "EUR")
    builder.domesticCurveBlock_ = usd_block
    builder.foreignCurveBlock_ = eur_block
    builder.fxSpot_ = 1.10
    builder.tolerance_ = 1.0e-8
    builder.solveMode_ = dal.CurveSolveMode.APPROXIMATE
    builder.initialGuess_ = 0.01
    builder.instruments_ = instruments
    builder.knotDates_ = knot_dates

    spec = builder.Build()
    result = dal.CalibrateXccyMarket(spec)

    assert result.market_ is not None  # nosec B101 - pytest assertions are intentional
    assert result.fxForwardCurve_ is not None  # nosec B101 - pytest assertions are intentional
    assert result.diagnostics_ is not None  # nosec B101 - pytest assertions are intentional


def test_staged_xccy_options_overload_preserves_default_results_and_aliases():
    """The additive overload retains the legacy call and exposes both naming styles."""
    spec, knot_dates = _make_xccy_spec(dal.CurveSolveMode.APPROXIMATE)
    legacy = dal.CalibrateXccyMarket(spec)
    options = dal.CrossCurrencyCalibrationOptions_()
    explicit = dal.CalibrateXccyMarket(spec, options)

    assert options.jacobianMode_ == dal.CurveJacobianMode.ANALYTIC  # nosec B101
    options.jacobian_mode = dal.CurveJacobianMode.BUMPED
    options.computeEffJacobianInverse_ = False
    options.compute_forward_jacobian = False
    assert options.jacobianMode_ == dal.CurveJacobianMode.BUMPED  # nosec B101
    assert not options.compute_eff_jacobian_inverse  # nosec B101
    assert not options.computeForwardJacobian_  # nosec B101

    assert legacy.diagnostics.market_rates == explicit.diagnostics_.marketRates_  # nosec B101
    assert legacy.diagnostics.model_rates == explicit.diagnostics_.modelRates_  # nosec B101
    assert legacy.diagnostics.residuals == explicit.diagnostics_.residuals_  # nosec B101
    assert legacy.fx_forward_curve.forwards == explicit.fxForwardCurve_.forwards_  # nosec B101
    assert explicit.diagnostics.parameter_knot_dates == knot_dates  # nosec B101
    assert explicit.diagnostics.parameterKnotDates_ == explicit.diagnostics.parameter_knot_dates  # nosec B101
    assert explicit.diagnostics.instrumentNames_ == explicit.diagnostics.instrument_names  # nosec B101
    assert explicit.diagnostics.residualTolerance_ == explicit.diagnostics.residual_tolerance  # nosec B101
    assert explicit.diagnostics.jacobianScaling_ == "unscaled"  # nosec B101
    assert explicit.diagnostics.jacobian_scaling == "unscaled"  # nosec B101
    assert explicit.diagnostics.effJacobianInverseScaling_ == "solver_scaled"  # nosec B101
    assert explicit.diagnostics.eff_jacobian_inverse_scaling == "solver_scaled"  # nosec B101
    assert explicit.diagnostics.jacobianAvailability_ == "not_available_for_mode"  # nosec B101
    assert explicit.diagnostics.eff_jacobian_inverse_availability == "not_available_for_mode"  # nosec B101
    assert explicit.diagnostics.jacobian_.rows() == 0  # nosec B101
    assert explicit.diagnostics.jacobian.cols() == 0  # nosec B101
    assert explicit.diagnostics.effJacobianInverse_.rows() == 0  # nosec B101
    assert explicit.diagnostics.eff_jacobian_inverse.cols() == 0  # nosec B101


def test_staged_xccy_exact_matrices_keep_diagnostics_ownership_and_axes():
    """Exact analytic matrices remain on staged diagnostics with reversed axes."""
    spec, knot_dates = _make_xccy_spec(dal.CurveSolveMode.EXACT, years=(2,))
    result = dal.CalibrateXccyMarket(spec, dal.CrossCurrencyCalibrationOptions_())
    diagnostics = result.diagnostics

    assert diagnostics.jacobian_availability == "available"  # nosec B101
    assert diagnostics.effJacobianInverseAvailability_ == "available"  # nosec B101
    assert diagnostics.jacobian.rows() == len(diagnostics.instrument_names)  # nosec B101
    assert diagnostics.jacobian_.cols() == len(knot_dates)  # nosec B101
    assert diagnostics.eff_jacobian_inverse.rows() == len(knot_dates)  # nosec B101
    assert diagnostics.effJacobianInverse_.cols() == len(diagnostics.instrumentNames_)  # nosec B101
    assert not hasattr(result, "jacobian_at_solution")  # nosec B101

File: dal-public/tests/test_xccy_calibration.cpp
//
// Created by wegamekinglc on 2026/6/20.
//

#include <gtest/gtest.h>

#include <dal-public/src/curvedata.hpp>
#include <dal-public/src/curveinstrument.hpp>
#include <dal-public/src/curveprotocol.hpp>
#include <dal-public/src/xccycalibration.hpp>

using Dal::CalibrateXccyMarket;
using Dal::CollateralType_OIS;
using Dal::CrossCurrencyCalibrationSpecBuilder_;
using Dal::CrossCurrencySwapNew;
using Dal::CurrencyPair_New;
using Dal::CurveBlockNew;
using Dal::CurveSolveMode_;
using Dal::Date_;
using Dal::DateTime_;
using Dal::DayBasis_New;
using Dal::DiscountPWLFNew;
using Dal::JointCurveDeclaration_;
using Dal::JointXccyCalibrationOptions_;
using Dal::JointXccyCalibrationResult_;
using Dal::JointXccyCalibrationSpecBuilder_;
using Dal::MarketFixingSnapshotNew;
using Dal::PeriodLength_New;
using Dal::RateIndexConvention_;
using Dal::RateIndexConvention_New;
using Dal::RateLegConvention_;
using Dal::RateLegConvention_New;
using Dal::String_;
using Dal::Vector_;

namespace {

    Date_ Today() { return Date_(2025, 6, 20); }
    Date_ Spot() { return Today().AddDays(2); }

    Dal::RateLegConvention_ Fixed6M365F() { return RateLegConvention_New(PeriodLength_New("6M"), DayBasis_New("ACT_365F")); }

    Dal::RateLegConvention_ Fixed6M360() { return RateLegConvention_New(PeriodLength_New("6M"), DayBasis_New("ACT_360")); }

    Dal::RateIndexConvention_ Libor3M() { return RateIndexConvention_New(PeriodLength_New("3M"), DayBasis_New("ACT_360"), CollateralType_OIS()); }

    Dal::RateIndexConvention_ Euribor6M() { return RateIndexConvention_New(PeriodLength_New("6M"), DayBasis_New("ACT_360"), CollateralType_OIS()); }

    Dal::RateLegConvention_ FloatLeg(const char* tenor, const char* basis) {
        return RateLegConvention_New(PeriodLength_New(tenor), DayBasis_New(basis));
    }

} // namespace

// CrossCurrencyCalibrationSpecBuilder_ defaults

TEST(XccyCalibrationTest, TestBuilderDefaults) {
    CrossCurrencyCalibrationSpecBuilder_ builder;
    ASSERT_NEAR(builder.tolerance_, 1.0e-10, 1e-15);
    ASSERT_NEAR(builder.fitTolerance_, 1.0e-6, 1e-15);
    ASSERT_EQ(builder.maxEvaluations_, 200);
    ASSERT_EQ(builder.maxRestarts_, 20);
    ASSERT_NEAR(builder.initialGuess_, 0.0, 1e-15);
    ASSERT_NEAR(builder.smoothingWeight_, 1.0, 1e-15);
    ASSERT_NEAR(builder.fxSpot_, 0.0, 1e-15);
    ASSERT_EQ(builder.solveMode_.Switch(), CurveSolveMode_::Value_::EXACT);
}

TEST(XccyCalibrationTest, TestJointBuilderDefaultsMatchCoreJointCalibration) {
    JointXccyCalibrationSpecBuilder_ builder;

    ASSERT_NEAR(builder.solverOptions_.tolerance_, 1.0e-8, 1.0e-15);
    ASSERT_NEAR(builder.solverOptions_.fitTolerance_, 1.0e-6, 1.0e-15);
    ASSERT_NEAR(builder.solverOptions_.initialGuess_, 0.0, 1.0e-15);
    ASSERT_EQ(builder.solverOptions_.maxEvaluations_, 200);
    ASSERT_EQ(builder.solverOptions_.maxRestarts_, 20);
    ASSERT_EQ(builder.solverOptions_.solveMode_.Switch(), CurveSolveMode_::Value_::EXACT);
}

// Build baseline curves for XCCY calibration

namespace {

    struct BaselineCurves_ {
        Dal::Handle_<Dal::CurveBlock_> domesticBlock_;
        Dal::Handle_<Dal::CurveBlock_> foreignBlock_;
    };

    BaselineCurves_ MakeBaselineCurves() {
        Vector_<Date_> knotDates;
        knotDates.push_back(Spot());
        knotDates.push_back(Spot().AddDays(3650)); // 10 years
        Vector_<> oisRates;
        oisRates.push_back(0.04);
        oisRates.push_back(0.04);

        auto usdOis = DiscountPWLFNew(String_("usd_ois"), String_("USD"), knotDates, oisRates);
        auto eurOis = DiscountPWLFNew(String_("eur_ois"), String_("EUR"), knotDates, oisRates);

        auto usdBlock = CurveBlockNew(usdOis, DayBasis_New("ACT_365F"));
        auto eurBlock = CurveBlockNew(eurOis, DayBasis_New("ACT_360"));

        BaselineCurves_ result;
        result.domesticBlock_ = usdBlock;
        result.foreignBlock_ = eurBlock;
        return result;
    }

} // namespace

// Cross-currency calibration

TEST(XccyCalibrationTest, TestCalibrateXccyMarket) {
    auto curves = MakeBaselineCurves();

    CrossCurrencyCalibrationSpecBuilder_ builder;
    builder.today_ = Today();
    builder.basisPair_ = CurrencyPair_New("USD", "EUR");
    builder.domesticCurveBlock_ = curves.domesticBlock_;
    builder.foreignCurveBlock_ = curves.foreignBlock_;
    builder.fxSpot_ = 1.10;
    builder.tolerance_ = 1.0e-8;
    builder.solveMode_ = CurveSolveMode_::Value_::APPROXIMATE;
    builder.initialGuess_ = 0.01;

    // Cross-currency swaps: USD fixed vs EUR float + spread
    auto usdLeg = Fixed6M365F();
    auto usdIndex = Libor3M();
    auto eurLeg = FloatLeg("6M", "ACT_360");
    auto eurIndex = Euribor6M();
    auto currencies = CurrencyPair_New("USD", "EUR");

    Vector_<Date_> knotDates;
    for (int y : {2, 5, 10}) {
        Date_ maturity = Spot().AddDays(y * 365);
        knotDates.push_back(maturity);
        builder.instruments_.push_back(
            CrossCurrencySwapNew(Today(), Spot(), maturity, 0.01, currencies, 100.0, 100.0 / 1.10, usdLeg, usdIndex, eurLeg, eurIndex));
    }
    builder.knotDates_ = knotDates;

    auto spec = builder.Build();
    auto result = CalibrateXccyMarket(spec);

    ASSERT_GT(result.diagnostics_.marketRates_.size(), static_cast<size_t>(0));
    ASSERT_EQ(result.diagnostics_.marketRates_.size(), result.diagnostics_.modelRates_.size());
    ASSERT_LT(result.diagnostics_.maxAbsResidual_, 1.0e-4);
    ASSERT_LT(result.diagnostics_.rmsResidual_, 1.0e-4);

    // Verify FX forward curve is populated
    ASSERT_GT(result.fxForwardCurve_.dates_.size(), static_cast<size_t>(0));
    ASSERT_EQ(result.fxForwardCurve_.dates_.size(), result.fxForwardCurve_.forwards_.size());
}

// Round-trip every builder field through Build() to guard against brace-init
// order/count drift between the builder and CrossCurrencyCalibrationSpec_.

TEST(XccyCalibrationTest, TestBuildRoundTripsEveryField) {
    auto curves = MakeBaselineCurves();

    CrossCurrencyCalibrationSpecBuilder_ b;
    b.today_ = Today();
    b.valuationTime_ = DateTime_(Today(), 9, 45);
    b.collateralCurrency_ = Dal::Ccy_("USD");
    b.fixings_ = MarketFixingSnapshotNew({{"USD-SOFR-3M", {{DateTime_(Today().AddDays(-1), 11, 0), 0.0425}}}});
    b.basisPair_ = CurrencyPair_New("USD", "EUR");
    b.domesticCurveBlock_ = curves.domesticBlock_;
    b.foreignCurveBlock_ = curves.foreignBlock_;
    b.fxSpot_ = 1.0825;
    b.fxForwardCollateral_ = CollateralType_OIS();
    b.smoothingWeight_ = 1.25;
    b.tolerance_ = 5.0e-11;
    b.fitTolerance_ = 6.0e-7;
    b.initialGuess_ = 0.0025;
    b.maxEvaluations_ = 233;
    b.maxRestarts_ = 17;
    b.solveMode_ = CurveSolveMode_::Value_::APPROXIMATE;
    b.knotDates_ = Vector_<Date_>{Spot(), Spot().AddDays(3650)};
    b.instruments_ = Vector_<Dal::Handle_<Dal::CrossCurrencySwap_>>{};

    auto spec = b.Build();

    ASSERT_EQ(spec.today_, b.today_);
    ASSERT_EQ(spec.valuationTime_, b.valuationTime_);
    ASSERT_EQ(spec.collateralCurrency_, b.collateralCurrency_);
    ASSERT_EQ(spec.fixings_.get(), b.fixings_.get());
    ASSERT_TRUE(spec.basisPair_ == b.basisPair_);
    ASSERT_EQ(spec.domesticCurveBlock_.get(), b.domesticCurveBlock_.get());
    ASSERT_EQ(spec.foreignCurveBlock_.get(), b.foreignCurveBlock_.get());
    ASSERT_NEAR(spec.fxSpot_, 1.0825, 1e-15);
    ASSERT_EQ(spec.fxForwardCollateral_.Switch(), b.fxForwardCollateral_.Switch());
    ASSERT_NEAR(spec.smoothingWeight_, 1.25, 1e-15);
    ASSERT_NEAR(spec.tolerance_, 5.0e-11, 1e-15);
    ASSERT_NEAR(spec.fitTolerance_, 6.0e-7, 1e-15);
    ASSERT_NEAR(spec.initialGuess_, 0.0025, 1e-15);
    ASSERT_EQ(spec.maxEvaluations_, 233);
    ASSERT_EQ(spec.maxRestarts_, 17);
    ASSERT_EQ(spec.solveMode_.Switch(), CurveSolveMode_::Value_::APPROXIMATE);
    ASSERT_EQ(spec.knotDates_.size(), static_cast<size_t>(2));
    ASSERT_TRUE(spec.instruments_.empty());
}

TEST(XccyCalibrationTest, TestJointBuilderRoundTripsEveryField) {
    const DateTime_ valuationTime(Today(), 9, 45);
    const DateTime_ fixingTime(Today().AddDays(-1), 11, 0);
    const auto fixings = MarketFixingSnapshotNew({{"USD-SOFR-3M", {{fixingTime, 0.0425}}}});

    JointCurveDeclaration_ domesticDiscount;
    domesticDiscount.curveName_ = "usd_ois";
    domesticDiscount.instruments_.push_back(DepositNew(Today(), Spot(), Spot().AddDays(30), 0.04, Libor3M()));
    domesticDiscount.knotDates_ = {Spot().AddDays(30), Spot().AddDays(365)};
    domesticDiscount.targetCollateral_ = CollateralType_OIS();
    domesticDiscount.targetTenor_ = PeriodLength_New("1M");
    domesticDiscount.calibrateDiscountCurve_ = true;
    domesticDiscount.baseLayeredOverDiscount_ = false;
    domesticDiscount.parameterization_ = Dal::CurveParameterization_::Value_::PIECEWISE_CONSTANT_FWD;
    domesticDiscount.logDfScheme_ = Dal::LogDfScheme_::Value_::LOG_CUBIC_NATURAL;
    domesticDiscount.smoothingWeight_ = 1.25;
    domesticDiscount.initialGuessPerNode_ = {0.01, 0.02};

    JointCurveDeclaration_ foreignForward;
    foreignForward.curveName_ = "eur_6m";
    foreignForward.instruments_.push_back(DepositNew(Today(), Spot(), Spot().AddDays(180), 0.03, Euribor6M()));
    foreignForward.knotDates_ = {Spot().AddDays(180), Spot().AddDays(730)};
    foreignForward.targetCollateral_ = CollateralType_OIS();
    foreignForward.targetTenor_ = PeriodLength_New("6M");
    foreignForward.calibrateDiscountCurve_ = false;
    foreignForward.baseLayeredOverDiscount_ = true;
    foreignForward.parameterization_ = Dal::CurveParameterization_::Value_::ZERO_RATE;
    foreignForward.logDfScheme_ = Dal::LogDfScheme_::Value_::LOG_LINEAR;
    foreignForward.smoothingWeight_ = 1.75;
    foreignForward.initialGuessPerNode_ = {0.03, 0.04};

    const auto xccy = CrossCurrencySwapNew(Today(), Spot(), Spot().AddDays(3650), 0.001, CurrencyPair_New("USD", "EUR"));

    JointXccyCalibrationSpecBuilder_ builder;
    builder.valuationTime_ = valuationTime;
    builder.pair_ = CurrencyPair_New("USD", "EUR");
    builder.collateralCurrency_ = Dal::Ccy_("USD");
    builder.fxSpot_ = 1.0825;
    builder.domestic_.ccy_ = Dal::Ccy_("USD");
    builder.domestic_.liborBasis_ = DayBasis_New("ACT_360");
    builder.domestic_.curves_ = {domesticDiscount};
    builder.foreign_.ccy_ = Dal::Ccy_("EUR");
    builder.foreign_.liborBasis_ = DayBasis_New("ACT_365F");
    builder.foreign_.curves_ = {foreignForward};
    builder.basis_.curveName_ = "usd_eur_basis";
    builder.basis_.instruments_ = {xccy};
    builder.basis_.knotDates_ = {Spot().AddDays(3650)};
    builder.basis_.parameterization_ = Dal::CurveParameterization_::Value_::PIECEWISE_CONSTANT_FWD;
    builder.basis_.smoothingWeight_ = 2.25;
    builder.basis_.initialGuessPerNode_ = {0.0025};
    builder.fixings_ = fixings;
    builder.solverOptions_.smoothingWeight_ = 9.5;
    builder.solverOptions_.tolerance_ = 5.0e-11;
    builder.solverOptions_.fitTolerance_ = 6.0e-7;
    builder.solverOptions_.initialGuess_ = 0.0075;
    builder.solverOptions_.maxEvaluations_ = 233;
    builder.solverOptions_.maxRestarts_ = 17;
    builder.solverOptions_.solveMode_ = CurveSolveMode_::Value_::APPROXIMATE;

    const auto spec = builder.Build();

    ASSERT_EQ(spec.valuationTime_, valuationTime);
    ASSERT_TRUE(spec.pair_ == builder.pair_);
    ASSERT_EQ(spec.collateralCurrency_, builder.collateralCurrency_);
    ASSERT_NEAR(spec.fxSpot_, 1.0825, 1.0e-15);
    ASSERT_EQ(spec.domestic_.ccy_, builder.domestic_.ccy_);
    ASSERT_EQ(spec.domestic_.liborBasis_.String(), builder.domestic_.liborBasis_.String());
    ASSERT_EQ(spec.domestic_.curves_.size(), static_cast<size_t>(1));
    ASSERT_EQ(spec.domestic_.curves_.front().curveName_, domesticDiscount.curveName_);
    ASSERT_EQ(spec.domestic_.curves_.front().instruments_.front().get(), domesticDiscount.instruments_.front().get());
    ASSERT_EQ(spec.domestic_.curves_.front().knotDates_, domesticDiscount.knotDates_);
    ASSERT_EQ(spec.domestic_.curves_.front().targetCollateral_, domesticDiscount.targetCollateral_);
    ASSERT_EQ(spec.domestic_.curves_.front().targetTenor_, domesticDiscount.targetTenor_);
    ASSERT_EQ(spec.domestic_.curves_.front().calibrateDiscountCurve_, domesticDiscount.calibrateDiscountCurve_);
    ASSERT_EQ(spec.domestic_.curves_.front().baseLayeredOverDiscount_, domesticDiscount.baseLayeredOverDiscount_);
    ASSERT_EQ(spec.domestic_.curves_.front().parameterization_, domesticDiscount.parameterization_);
    ASSERT_EQ(spec.domestic_.curves_.front().logDfScheme_, domesticDiscount.logDfScheme_);
    ASSERT_NEAR(spec.domestic_.curves_.front().smoothingWeight_, domesticDiscount.smoothingWeight_, 1.0e-15);
    ASSERT_EQ(spec.domestic_.curves_.front().initialGuessPerNode_, domesticDiscount.initialGuessPerNode_);
    ASSERT_EQ(spec.foreign_.ccy_, builder.foreign_.ccy_);
    ASSERT_EQ(spec.foreign_.liborBasis_.String(), builder.foreign_.liborBasis_.String());
    ASSERT_EQ(spec.foreign_.curves_.size(), static_cast<size_t>(1));
    ASSERT_EQ(spec.foreign_.curves_.front().curveName_, foreignForward.curveName_);
    ASSERT_EQ(spec.foreign_.curves_.front().instruments_.front().get(), foreignForward.instruments_.front().get());
    ASSERT_EQ(spec.foreign_.curves_.front().knotDates_, foreignForward.knotDates_);
    ASSERT_EQ(spec.foreign_.curves_.front().targetCollateral_, foreignForward.targetCollateral_);
    ASSERT_EQ(spec.foreign_.curves_.front().targetTenor_, foreignForward.targetTenor_);
    ASSERT_EQ(spec.foreign_.curves_.front().calibrateDiscountCurve_, foreignForward.calibrateDiscountCurve_);
    ASSERT_EQ(spec.foreign_.curves_.front().baseLayeredOverDiscount_, foreignForward.baseLayeredOverDiscount_);
    ASSERT_EQ(spec.foreign_.curves_.front().parameterization_, foreignForward.parameterization_);
    ASSERT_EQ(spec.foreign_.curves_.front().logDfScheme_, foreignForward.logDfScheme_);
    ASSERT_NEAR(spec.foreign_.curves_.front().smoothingWeight_, foreignForward.smoothingWeight_, 1.0e-15);
    ASSERT_EQ(spec.foreign_.curves_.front().initialGuessPerNode_, foreignForward.initialGuessPerNode_);
    ASSERT_EQ(spec.basis_.curveName_, builder.basis_.curveName_);
    ASSERT_EQ(spec.basis_.instruments_.front().get(), xccy.get());
    ASSERT_EQ(spec.basis_.knotDates_, builder.basis_.knotDates_);
    ASSERT_EQ(spec.basis_.parameterization_, builder.basis_.parameterization_);
    ASSERT_NEAR(spec.basis_.smoothingWeight_, 2.25, 1.0e-15);
    ASSERT_EQ(spec.basis_.initialGuessPerNode_, builder.basis_.initialGuessPerNode_);
    ASSERT_EQ(spec.fixings_.get(), fixings.get());
    ASSERT_NEAR(spec.tolerance_, 5.0e-11, 1.0e-15);
    ASSERT_NEAR(spec.fitTolerance_, 6.0e-7, 1.0e-15);
    ASSERT_NEAR(spec.initialGuess_, 0.0075, 1.0e-15);
    ASSERT_EQ(spec.maxEvaluations_, 233);
    ASSERT_EQ(spec.maxRestarts_, 17);
    ASSERT_EQ(spec.solveMode_.Switch(), CurveSolveMode_::Value_::APPROXIMATE);
}

TEST(XccyCalibrationTest, TestJointOptionsAndResultLayoutArePublic) {
    JointXccyCalibrationOptions_ options;
    options.jacobianMode_ = Dal::CurveJacobianMode_::Value_::BUMPED;
    options.computeEffJacobianInverse_ = false;
    options.computeForwardJacobian_ = false;
    using CalibrateWithOptions_ = JointXccyCalibrationResult_ (*)(const Dal::JointXccyCalibrationSpec_&, const JointXccyCalibrationOptions_&);
    const CalibrateWithOptions_ calibrate = static_cast<CalibrateWithOptions_>(&Dal::CalibrateJointXccyMarket);

    ASSERT_NE(calibrate, nullptr);
    ASSERT_EQ(options.jacobianMode_.Switch(), Dal::CurveJacobianMode_::Value_::BUMPED);
    ASSERT_FALSE(options.computeEffJacobianInverse_);
    ASSERT_FALSE(options.computeForwardJacobian_);

    JointXccyCalibrationResult_ result;
    result.parameterRanges_ = {{"domestic:usd_ois", 0, 2}, {"basis:usd_eur", 2, 1}};
    result.residualRanges_ = {{"domestic:usd_ois", 0, 2}, {"xccy:usd_eur", 2, 1}};
    result.marketRates_ = {0.04, 0.041, 0.001};
    result.modelRates_ = {0.04, 0.041, 0.001};
    result.residuals_ = {0.0, 0.0, 0.0};

    ASSERT_EQ(result.parameterRanges_.size(), static_cast<size_t>(2));
    ASSERT_EQ(result.parameterRanges_[1].name_, Dal::String_("basis:usd_eur"));
    ASSERT_EQ(result.parameterRanges_[1].offset_, 2);
    ASSERT_EQ(result.parameterRanges_[1].size_, 1);
    ASSERT_EQ(result.residualRanges_.size(), static_cast<size_t>(2));
    ASSERT_EQ(result.residualRanges_[1].name_, Dal::String_("xccy:usd_eur"));
    ASSERT_EQ(result.marketRates_.size(), static_cast<size_t>(3));
    ASSERT_EQ(result.modelRates_.size(), static_cast<size_t>(3));
    ASSERT_EQ(result.residuals_.size(), static_cast<size_t>(3));
}
